### PR TITLE
support hosts

### LINKED ISSUE #2150: 自定义hosts的功能
我知道这是一个已经存在的问题 #544 但3年前的实现和现在ss-android的实现有些不同。
我用的购买的ss服务，因此没有ss服务端的权限，ss服务端设了超时比较短，60秒没数据就断开连接，于是fcm就频繁的断开连接。我自己用golang写了一个fcm的中转代理，在国内和国外的vps上都搭了一个针对5228端口的转发，速度不快，但可以满足fcm push的稳定性需要。
以前是通过magisk的systemless hosts功能来修改hosts，把fcm涉及的多个域名全部指向我的服务器来提供服务的。
但更换了华为P20后，无法解锁也没有root能力，无法修改hosts，这条路走不通了。
看了现在的ss-android的方案，之前用的overture是可以指定hosts文件的但也是写死的，最新的用kotlin重写的dns server没有这个能力。
我想加入这个功能但不提供界面，通过targetApi>=19时尝试读取Context.getExternalFilesDir()下的hosts文件来提供这个能力（不存在或无权限则忽略），这样也不需要额外申请权限。
不知道做出这个功能后是否有可能合入？

### PR DESCRIPTION
**Pull request by ayanamist**
fix #2150

### CODE REVIEW

**Comment by Mygod** [comment]
In our use case, parseNumericAddress is performance critical as it needs to be invoked once for every rule in ACL. That's why we are using inet_pton and parseNumericAddress as a fallback.

You can attach some benchmark for `Inetaddresses` or keep that in a separate pull request.

**Comment by ayanamist** [comment]
@Mygod 我这个是给解析hosts文件用的，不可能单独剥离出来。另外实现是从guava里抠的（代码里注释写了port from），不是我自己写的。如果你们不用，那就用另外的实现好了。
目前hosts的实现已经基本ready了，我本来是想着周末测测再推上来的，只是刚好看到你们有类似的需求。

**Comment by Mygod** [comment]
Use this for now: https://github.com/shadowsocks/shadowsocks-android/blob/6b60d91b4be93bff7bee04dede031a622772ea84/core/src/main/java/com/github/shadowsocks/utils/Utils.kt#L57-L58

**Comment by ayanamist** [comment]
> Use this for now:
> 
> [shadowsocks-android/core/src/main/java/com/github/shadowsocks/utils/Utils.kt](https://github.com/shadowsocks/shadowsocks-android/blob/6b60d91b4be93bff7bee04dede031a622772ea84/core/src/main/java/com/github/shadowsocks/utils/Utils.kt#L57-L58)
> 
> Lines 57 to 58 in [6b60d91](/shadowsocks/shadowsocks-android/commit/6b60d91b4be93bff7bee04dede031a622772ea84)
> 
>  fun String?.parseNumericAddress(): InetAddress? = Os.inet_pton(OsConstants.AF_INET, this) 
>          ?: Os.inet_pton(OsConstants.AF_INET6, this)?.let { parseNumericAddress.invoke(null, this) as InetAddress }

已经改了。

File: mobile/build.gradle [hunk 1]
```diff
@@ -40,14 +40,8 @@ android {
...
-    splits {
-        abi {
-            enable true
-            universalApk true
-        }
-    }
-    sourceSets.main.jniLibs.srcDirs +=
-            new File(project(':core').buildDir, "intermediates/bundles/${getCurrentFlavor()}/jni")
+    sourceSets.main.jni.srcDirs = []
+    sourceSets.main.jniLibs.srcDirs = []
```
Comment (Mygod) [inline mobile/build.gradle]: Please open a separate pull request if you want to suggest this change.

File: core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt [hunk 2]
```diff
@@ -54,6 +57,8 @@ class LocalDnsServer(private val localResolver: suspend (String) -> Array<InetAd
     var tcp = true
     var remoteDomainMatcher: Regex? = null
     var localIpMatcher: List<Subnet> = emptyList()
+    @Volatile
+    var hostsMap: Map<String, List<InetAddress>> = HashMap()
```
Comment (Mygod) [inline core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt]: Please remove `@Volatile var` and initialize hostsMap in constructor instead.

File: core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt [hunk 3]
```diff
@@ -102,10 +121,15 @@ class LocalDnsServer(private val localResolver: suspend (String) -> Array<InetAd
         return supervisorScope {
             val remote = async { withTimeout(TIMEOUT) { forward(packet) } }
             try {
-                if (forwardOnly || request.header.opcode != Opcode.QUERY) return@supervisorScope remote.await()
+                if (request.header.opcode != Opcode.QUERY) return@supervisorScope remote.await()
                 val question = request.question
                 if (question?.type != Type.A) return@supervisorScope remote.await()
                 val host = question.name.toString(true)
+                val hostsResults = hostsResolver(host)
+                if (hostsResults.isNotEmpty()) {
```
Comment (Mygod) [inline core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt]: `remote.cancel()` missing.

File: core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt [hunk 4]
```diff
@@ -142,6 +160,41 @@ class LocalDnsServer(private val localResolver: suspend (String) -> Array<InetAd
...
+        try {
+            val hostsFile = File(Core.deviceStorage.getExternalFilesDir(null), "hosts")
+            hostsFile.createNewFile()
+            hostsFile.forEachLine {
+                val line = it.substringBefore('#')
+                if (line.isEmpty()) {
+                    return@forEachLine
+                }
+
+                val splitted = line.split(hostsDelimiter)
```
Comment (Mygod) [inline core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt]: `split` should ignore empty entries.

File: core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt [hunk 5]
```diff
@@ -142,6 +160,41 @@ class LocalDnsServer(private val localResolver: suspend (String) -> Array<InetAd
...
+                if (splitted.size < 2) {
+                    return@forEachLine
+                }
+                val addr = splitted[0].parseNumericAddress() ?: return@forEachLine
+                for (j in 1 until splitted.size) {
```
Comment (Mygod) [inline core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt]: It's better to iterate over `splitted.skip(1)`.

File: mobile/build.gradle (hunk 1 above)
Comment (ayanamist) [reply mobile/build.gradle]: 不好意思，这是我调试时的东西，不小心带上来了，我去掉。

**Comment by ayanamist** [comment]
> This seems to be missing STORAGE permission but it doesn't matter. Also [here](https://developer.android.com/preview/privacy/scoped-storage)'s why I don't support using external storage.

我是看到这里 https://developer.android.com/reference/android/content/Context.html#getExternalFilesDir(java.lang.String) 提到的
> Starting in Build.VERSION_CODES.KITKAT, no permissions are required to read or write to the returned path; it's always accessible to the calling app.
所以特意没有声明权限。因为这个目录在sdcard里，Q之前，可以用Solid Explorer等编辑器来直接修改这个文件，类似root后的手机直接修改`/etc/hosts`一样。这也是因为我没有写UI代码能力的局限之处。如果有UI了，可以不用这样做。

File: core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt (hunk 4 above)
Comment (ayanamist) [reply core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt]: 我下面有`splitted.size < 2`的判断，这里还需要显式再判断吗？我这段逻辑也是port自https://golang.org/src/net/hosts.go 这里面的parse也没判断空行

File: core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt (hunk 5 above)
Comment (ayanamist) [reply core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt]: splitted对应的是List<String>，是没有skip这个方法的，最接近的是drop，但drop的实现，是要把数据全部复制一遍的，就是有1w条记录，就把剩下9999条复制到一个新的list再返回。这样性能非常差吧。毕竟肯定会有人用hosts功能来做ad block的。现在这个设计也是摘自golang的实现，可读性稍差，但性能不会有问题。

File: core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt (hunk 4 above)
Comment (Mygod) [reply core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt]: I mean you should remove leading empty spaces, for example would your parser work properly for `" 127.0.0.1 localhost "`?

File: core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt (hunk 5 above)
Comment (Mygod) [reply core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt]: Use `.asStream()` if you concern about performance. I personally wouldn't worry about that.

Comment (ayanamist) [reply core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt]: `.stream()`要求api level 24……

Comment (Mygod) [reply core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt]: *`asSequnce()`

File: core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt (hunk 4 above)
Comment (ayanamist) [reply core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt]: OK，这里加了trim，另外看了下golang的实现，用一个很巧妙的方式做了trim left的操作，是我看漏了。

**Comment by madeye** [comment]
Can we follow the profile import logic to import hosts file?

**Comment by Mygod** [comment]
I will make it a global option.

**Comment by ayanamist** [comment]
> Can we follow the profile import logic to import hosts file?

这好像是ui部分的改动？这个我无能为力了

**Comment by Mygod** [comment]
@ayanamist Thank you for your contribution. Please review the code. And before you ask, yes I did expect you to write these UI codes. :smile: 

I had to rebase this onto `preference-1.1` branch (see #2162), which sadly will take a while before it can be merged, but I see you are compiling this repo by yourself so it should not be an issue for you.

I have not tested the hosts functionality yet, you can look into that. Also you are welcome to contribute a similar UI for TV (if not I will do it later).

**Comment by ayanamist** [comment]
我看现在是用merge的方式把代码从`preference-1.1`分支合并过来，需要我用rebase的方式重新弄一下吗？另外ui部分我确实搞不定，完全没有android ui知识……

**Comment by Mygod** [comment]
No everything is fine now. Just give this ui a spin and see if it works properly.

**Comment by Mygod** [comment]
@madeye @ayanamist Please review the changes and do some testing.

**Comment by madeye** [comment]
Yeah, I will check this later.

**Comment by madeye** [comment]
Cool! Test it locally and works very well.

**Comment by Mygod** [comment]
@ayanamist Thoughts? If not, we will merge this into #2162.

**Comment by ayanamist** [comment]
> @ayanamist Thoughts? If not, we will merge this into #2162.

LGTM

**Pull request merged**